src/core/internal/array/equality.d

```diff
@@ -23,20 +23,50 @@ bool __ArrayEq(T1, T2)(T1[] a, T2[] b)
     return true;
 }
 
+ // `lhs == rhs` lowers to `__equals(lhs, rhs)` for dynamic arrays
+ // The scalar-only overload takes advantage of known properties of scalars to
+ // consolidate IFTI instances and remove the need for helper functions.
+bool __equals(T1, T2)(scope const T1[] lhs, scope const T2[] rhs)
+@nogc nothrow pure @trusted
+if (__traits(isScalar, T1) && __traits(isScalar, T2))
+{
+    if (lhs.length != rhs.length)
+        return false;
+    static if (T1.sizeof == T2.sizeof
+        && __traits(isUnsigned, T1) == __traits(isUnsigned, T2)
+        && !__traits(isFloating, T1) && !__traits(isFloating, T2))
+    {
+        if (!__ctfe)
+        {
+            // This would improperly allow equality of integers and pointers
+            // but the CTFE branch will stop this function from compiling then.
+            import core.stdc.string : memcmp;
+            return lhs.length == 0 || 0 == memcmp(lhs.ptr, rhs.ptr, lhs.length * T1.sizeof);
+        }
+    }
+    foreach (const i; 0 .. lhs.length)
+        if (lhs.ptr[i] != rhs.ptr[i])
+            return false;
+    return true;
+}
+
+pragma(inline, true)
+private @trusted ref R at(R)(return scope R[] r, size_t i) { return r.ptr[i]; }
```

Comment:
If I change the argument of `at` to be `inout` and its return type to be inferred (but still a reference) I get an error in a DMD test. It seems like that shouldn't happen so I am noting this as a reminder to maybe look into the cause later.

```console
Test compilable/interpret3.d failed.  The logged output:
/home/braddr/sandbox/at-client/pull-4107266-Linux_32_64/dmd/generated/linux/release/32/dmd -conf= -m64 -Icompilable  -fPIC  -odgenerated/compilable -ofgenerated/compilable/interpret3_0.o  -c compilable/interpret3.d 
compilable/interpret3.d(2914): Deprecation: `case` variables have to be `const` or `immutable`
compilable/interpret3.d(6313): Deprecation: identity comparison of static arrays implicitly coerces them to slices, which are compared by reference
/home/braddr/sandbox/at-client/pull-4107266-Linux_32_64/dmd/test/../../druntime/import/core/internal/array/equality.d(54): Error: reinterpreting cast from `immutable(string[2])*` to `inout(char[][2])*` is not supported in CTFE
/home/braddr/sandbox/at-client/pull-4107266-Linux_32_64/dmd/test/../../druntime/import/core/internal/array/equality.d(92):        called from here: `at(lhs, u)`
```